lib/spack/spack/spec.py

```diff
@@ -985,6 +985,7 @@ def __init__(self, spec_like=None,
         self._dependents = DependencyMap()
         self._dependencies = DependencyMap()
         self.namespace = None
+        self._explicit_providers = {}
```

Comment:
I don't think this should be a spec-level attribute.  It should be an *edge* attribute in the DAG -- and I think it should also be parsed as one.

Reply:
I'm reading the code that does concretization. This somehow serves the same purpose as:
```python
user_spec_deps = self.flat_dependencies(copy=False)
```
in the sense that it contains the user specified edge attributes. What I can do is construct this dictionary by analyzing the flat DAG passed as input to concretization and then pass it around to helper function as we do for `user_spec_deps`. Does that seem sensible?